### PR TITLE
Add gather

### PR DESCRIPTION
**Pull request by tkf**
Hi, thanks for implementing this library!

I wanted to play with the gather operation so I implemented it using SIMD.jl infrastructure (whish was surprisingly straightforward).  Do you think it's good to add this?

As a side note, if you want to get fancy, I think indexing `arr[idx]` by a `Vec` can be a sugar for `vgather` (and `arr[idx] = vec` for scatter in the future).

### CODE REVIEW

**Comment by eschnett** [comment]
@tkf Gather or scatter operations sound very interesting. I'd be happy to add them.

I haven't studied your code in detail yet, but it would need to have:
- test cases
- a benchmark (or disassembled code) that shows it is efficient, i.e. it should use hardware gather SIMD instructions where possible, and should generate reasonable code otherwise

The ("fancy") indexing syntax you propose looks good as well.

Thanks!

**Comment by tkf** [comment]
> * test cases

It has some tests.  In the latest commit I refactored/improved it a little bit more.

> * a benchmark (or disassembled code)

I did some benchmark with for the sparse-dense dot product using `vgather`.  Both a "straightforward" SIMD implementation (`dot_simd`) and a "pointer based" SIMD implementation (`dot_ptr`) outperform the scalar implementation (`dot_simple`).  In the beset case input size, they can be about twice as fast as than the scalar implementation:

benchmark

(Side note: I tried both aligned and non-aligned cases but I couldn't find the difference consistently.  The difference between them seems to be just a noise.)

Here is the implementation of `dot_simd`;

```julia
@inline function dot_simd(sv::SpVect, ys::AbstractVector,
                          ::Val{align} = Val(false),
                          ::Val{N} = Val(4),
                          ) where {N, align}
    Ti = eltype(sv.nzind)
    Tv = eltype(sv.nzval)
    T = promote_type(eltype(sv.nzval), eltype(ys))
    isempty(sv.nzval) && return zero(T)

    nz_size = length(sv.nzval)
    nomask = Vec(ntuple(_ -> true, N))
    vacc = zero(Vec{N, T})
    @inbounds for j in 1:N:(nz_size - N + 1)
        idx = vload(Vec{N, Ti}, sv.nzind, j, Val{align})
        bs = vgather(ys, idx, nomask, Val{align})
        as = vload(Vec{N, Tv}, sv.nzval, j, Val{align})
        vacc = muladd(as, bs, vacc)
    end

    acc = sum(vacc)
    @inbounds for j in (nz_size - nz_size % N + 1):nz_size
        acc = muladd(sv.nzval[j], ys[sv.nzind[j]], acc)
    end
    return acc
end
```

Looking at `code_native` output, `SIMD.vgather` indeed compiles to `vgatherqpd` in my laptop (which has [this CPU](https://ark.intel.com/products/122589/Intel-Core-i7-8550U-Processor-8M-Cache-up-to-4-00-GHz-)):

```asm
; Location: SIMD.jl:1490
; Function vgather; {
; Location: SIMD.jl:1447
; Function macro expansion; {
; Location: SIMD.jl:1477
        vpcmpeqd        %ymm4, %ymm4, %ymm4
        vxorpd  %xmm5, %xmm5, %xmm5
        vgatherqpd      %ymm4, (,%ymm3), %ymm5
;}}}
```

You can find all source code here: https://gist.github.com/926802a335ec491784192d8566a42cf1

`dot_simple` is the baseline scalar implementation.  `dot_ptr` is pointer-based implementation that uses SIMD.  This is to see if we can improve the setup cost and be faster than (or at least as fast as) `dot_simple` for all inputs.

**Comment by KristofferC** [comment]
> It looks like the overhead is in StepRange creation and especially calculating the remainder. U

Yes, non `UnitRanges` should be avoided in performance sensitive code until this is fixed, ref, https://github.com/JuliaLang/julia/issues/26770.

**Comment by tkf** [comment]
Thanks for the link.  Good to know it's already tracked in Julia issue.

**Comment by tkf** [comment]
@eschnett I go ahead and add scatter as well.  ATM, I don't have a benchmark showing that it is effective since I don't have a quick access to a machine with AVX-512 (which is required for hardware scatter IIUC).  Let me know if you want to wait until there is an effective example (so that I can exclude the commit).

File: README.md [hunk 1]
```diff
@@ -43,7 +43,7 @@ The SIMD package provides the usual arithmetic and logical operations for SIMD v
 
 (Currently missing: `count_ones count_zeros exponent ldexp leading_ones leading_zeros significand trailing_ones trailing_zeros`, many trigonometric functions)
 
-(Also currently missing: Type conversions, reinterpretation that changes the vector size, scatter/gather operations, masked load/store operations)
```
Comment (tkf) [inline README.md]: I removed "masked load/store operations" as well since it seems `vload` and `vstore` support mask already.

**Comment by eschnett** [comment]
@tkf Sorry for the wait. I think the current state is fine. Let me know whether I should merge this, and create a new release – I'm not quite clear about your intent.

**Comment by tkf** [comment]
I've done what I wanted to include in this RP. It would be nice if you can merge (and release).

**Comment by tkf** [comment]
Thanks a lot!

**Pull request merged**